File: DFS&BFS/javascript/12851.js
const [n, k] = require("fs").readFileSync("/dev/stdin").toString().split(" ").map(Number);

const checkCondition = (pos, visited) => {
  return pos >= 0 && pos <= 100000 && !visited[pos];
};

const solution = (n, k) => {
  const q = [];
  const visited = new Array(100002).fill(false);
  const answer = [0, 0];
  let idx = 0;
  q.push([n, 0]);

  if (n > k) {
    return [n - k, 1].join("\n");
  }

  while (q.length > idx) {
    const [pos, cnt] = q[idx++];
    visited[pos] = true;

    if (answer[0] !== 0 && cnt > answer[0]) {
      break;
    }
    if (pos === k) {
      answer[0] = cnt;
      answer[1]++;
    }

    if (checkCondition(pos - 1, visited)) {
      q.push([pos - 1, cnt + 1]);
    }

    if (checkCondition(pos + 1, visited)) {
      q.push([pos + 1, cnt + 1]);
    }

    if (checkCondition(pos * 2, visited)) {
      q.push([pos * 2, cnt + 1]);
    }
  }
  return answer.join("\n");
};

const answer = solution(n, k);
console.log(answer);

File: DFS&BFS/python/12851.py
import sys
from collections import deque

input = sys.stdin.readline
n,k = map(int,input().split())

def solution(n,k):
  q = deque()
  q.append([n,0])
  answer = [0,0]
  visited = [False for _ in range(100001)]

  if n>k:
    print("\n".join([str(n-k),"1"]))
    exit()

  while q:
    pos,cnt = q.popleft()
    visited[pos] = True

    if answer[0] != 0 and cnt > answer[0]:
      break

    if pos == k:
      answer[0] = cnt
      answer[1] += 1
    
    if answer[0] != 0:
      continue

    if pos<0 or pos>=100001:
      continue

    if 0 <= pos-1 <=100000 and not visited[pos-1] :
      q.append([pos-1,cnt+1])
    
    if 0 <= pos+1 <=100000 and not visited[pos+1] :
      q.append([pos+1,cnt+1])

    if 0 <= pos*2 <=100000 and not visited[pos*2] :
      q.append([pos*2,cnt+1])

  return "\n".join(map(str,answer))
answer = solution(n,k)
print(answer)
